lib/workers.js

```diff
@@ -0,0 +1,294 @@
+      options: simplifyObject(workerObject.options),
+      tests: workerObject.tests,
+      testRoot: workerObject.testRoot,
+      workerIndex: workerObject.workerIndex + 1,
+    },
+  });
+};
+
+const simplifyObject = (object) => {
+  return Object.keys(object)
+    .filter(k => k.indexOf('_') !== 0)
+    .filter(k => typeof object[k] !== 'function')
+    .filter(k => typeof object[k] !== 'object')
+    .reduce((obj, key) => {
+      obj[key] = object[key];
+      return obj;
+    }, {});
+};
+
+const repackTest = (test) => {
+  test = Object.assign(new Test(test.title || '', () => { }), test);
+  test.parent = Object.assign(new Suite(test.parent.title), test.parent);
+  return test;
+};
+
+const createWorkerObjects = (testGroups, config, testRoot) => {
+  return testGroups.map((tests, index) => {
+    const workerObj = new WorkerObject(index);
+    workerObj.addConfig(config);
+    workerObj._addTests(tests);
+    workerObj.setTestRoot(testRoot);
+
+    return workerObj;
+  });
+};
+
+const indexOfSmallestElement = (groups) => {
+  let i = 0;
+  for (let j = 1; j < groups.length; j++) {
+    if (groups[j - 1].length > groups[j].length) {
+      i = j;
+    }
+  }
+  return i;
+};
+
+const convertToMochaTests = (testGroup) => {
+  const group = [];
+  if (testGroup instanceof Array) {
+    const mocha = MochaFactory.create({}, {});
+    mocha.files = testGroup;
+    mocha.loadFiles();
+    mocha.suite.eachTest((test) => {
+      const { id } = test;
+      group.push(id);
+    });
+    mocha.unloadFiles();
+  }
+
+  return group;
+};
+
+class WorkerObject {
+  /**
+   * @param {Number} workerIndex - Unique ID for worker
+   */
+  constructor(workerIndex) {
+    this.workerIndex = workerIndex;
+    this.options = {};
+    this.tests = [];
+    this.testRoot = getTestRoot();
+  }
+
+  addConfig(config) {
+    const oldConfig = JSON.parse(this.options.override || '{}');
+    const newConfig = {
+      ...oldConfig,
+      ...config,
+    };
+    this.options.override = JSON.stringify(newConfig);
+  }
+
+  addTestFiles(testGroup) {
+    this._addTests(convertToMochaTests(testGroup));
+  }
+
+  _addTests(tests) {
+    this.tests = this.tests.concat(tests);
+  }
+
+  setTestRoot(path) {
+    this.testRoot = getTestRoot(path);
+  }
+}
+
+class Workers extends EventEmitter {
+  /**
+   * @param {Number} numberOfWorkers
+   * @param {Object} config
+   */
+  constructor(numberOfWorkers, config = { by: 'test' }) {
+    super();
+    this.codecept = initializeCodecept(config.testConfig);
+    this.finishedTests = [];
+    this.errors = [];
+    this.numberOfWorkers = 0;
+    this.closedWorkers = 0;
+    this.workers = [];
+    this.stats = {
+      passes: 0,
+      failures: 0,
+      tests: 0,
+    };
+    this.testGroups = [];
+
+    this._initWorkers(numberOfWorkers, config);
+  }
+
+  _initWorkers(numberOfWorkers, config) {
+    if (isFunction(config.by)) {
+      const createTests = config.by;
+      const testGroups = createTests();
```

Comment:
Mocha.suite should be passed into that function so we could iterate over tests

Reply:
@DavertMik We will be getting the file names as array from that function. I thought it will be easy for users to just send file paths rather than test id. 

```js
const testGroups = () => {

  const files = [
    ['/path/to/file/guthub_test.js', '/path/to/file/devto_test.js'],
    ['/path/to/file/devto_test.js']
  ];

  return files;
}
```

your view?